### PR TITLE
Fixed 500 Error - Attendee not found

### PR DESCRIPTION
**Pull request by timrxd**
This just puts up a simple page saying "500 error - Attendee not found" whenever it gets a 500 response.

### CODE REVIEW

**Comment by binary1230** [comment]
hey Tim, I think this might change every 500 error page in the system.

this is for https://github.com/magfest/ubersystem/issues/1564 ?

I think what you want to do instead is catch this upstream so it's just not an error in the first place.

in uber/uber/models.py , look for this line:

```
if "registration" in url:
                with Session() as session:
                    attendee = session.query(Attendee).filter(Attendee.id == params['id']).first()
                    Tracking.track(c.PAGE_VIEWED, attendee)
```

Attendee there might return ```None```, and you can do something like:

```
if "registration" in url:
                with Session() as session:
                    attendee = session.query(Attendee).filter(Attendee.id == params['id']).first()
                    if attendee:
                            Tracking.track(c.PAGE_VIEWED, attendee)
```

(bonus points: you can do the same thing with the group code right below that).

Basically, end-users see that 500 error page anytime programmers don't catch an error they're supposed to.  So the solution here is to catch the error in the first place before it generates a 500 server error.

**Comment by timrxd** [comment]
Now it only calls that page when attendee is None.  If that works I'll add the statement to the groups section.

**Comment by binary1230** [comment]
so, one note here, this is in our "tracking" code which tracks when people look at attendees.

there's a precedent right above this section that shows a check that just ignores the rest of this tracking if there's not a valid attendee id:

```
             # Only log the page view if there's a valid attendee ID
             params = dict(parse_qsl(query))
             if 'id' not in params or params['id'] == 'None':
                 return
```

that check doesn't handle the case where the id looks valid, but it's not (like 'ffffffffffffffff').  it probably should, but, rather than do another DB query to check if the page exists, we can just check further down in the tracking code, something like:

```
if "registration" in url:
                with Session() as session:
                    attendee = session.query(Attendee).filter(Attendee.id == params['id']).first()
                    if not attendee:
                          return
                    Tracking.track(c.PAGE_VIEWED, attendee)
```

that'll eliminate the need for a redirect and a custom error page, at least in the case of this tracking code, and bailing out here is already established as OK by the existing check higher up in this function

fixing this might still trigger a different 500 error in registration.py form(), so we'll see what happens.  that would be a second bug hidden by this one.

one other small thing: python's (picky) standards mandate that you have a newline at the end of each file, that's why the integration tests on github are complaining right now.

keep going! nearly there!
\m/
-Dom

File: uber/models.py [hunk 1]
```diff
@@ -1805,6 +1805,8 @@ def track_pageview(cls, url, query):
             if "registration" in url:
                 with Session() as session:
                     attendee = session.query(Attendee).filter(Attendee.id == params['id']).first()
+                    if (attendee == None):
+                        raise HTTPRedirect('../registration/error_attendee_not_found')
```
Comment (EliAndrewC) [inline uber/models.py]: Looks good! Two minor things about this code:

1) if-statements in Python don't need parenthesis.

2) Although ``if attendee == None`` is indeed valid, the idiomatic way to compare to ``None`` in Python is ``if attendee is None``.  That's just a style thing though, since both versions do basically the same thing in this case (the latter being slightly more efficient).

File: uber/site_sections/registration.py [hunk 2]
```diff
@@ -901,3 +901,6 @@ def inactive(self, session):
                                 .filter(~Attendee.badge_status.in_([c.NEW_STATUS, c.COMPLETED_STATUS]))
                                 .order_by(Attendee.badge_status, Attendee.full_name).all()
         }
+
+    def error_attendee_not_found(message):
+        return "Sorry, but that attendee doesn't exist."
```
Comment (EliAndrewC) [inline uber/site_sections/registration.py]: Two minor things:

1) The parameter to this function should be named ``self``, which is the Python version of ``this`` from other languages such as C / C++ / Java / Javascript / etc.  The first parameter to any instance method such as this will automatically be the class instance, which by convention you should always name ``self`` unless you have a really good reason not to.

2) As Dom already mentioned, our style checker requires newlines at the end of each file (which is what the PEP-8 style guide recommends), so you'll have to change your editor / IDE to do that.

Slightly more important question:

This page handler is currently in ``registration.py``, which is a page which is restricted to admins.  I'm not sure whether this is an error page we'll ever want to use for actions taken by non-admins, but if it is then we probably want it somewhere else where anyone can access it.  If we decide to do that, then ``preregistration.py`` is probably a good place for it.

**Comment by timrxd** [comment]
###### Looking into return instead of redirecting to the custom page:
If I do an empty return when the attendee is None, we get past that check fine, but then we end up in the _make_getter() method.  In models.py, line 1834:
```
inst = self.query(model).filter_by(id=id).one()
```
which raises NoResultFound: No Row Found.  I can catch that exception, but I'm not sure what to do to handle it.  If I return out of this one, I get a bunch of NoneType exceptions, and I don't think the code was designed to handle a None getter.  A few lines up, if the id is defined as None, it sets the inst to model(), which brings up a blank form to enter id=ffffffff-ffff-ffff-ffff-ffffffffffff's information.  I don't know if that should be fix, or if you want an error page to report that the id doesn't exist.  


Also, next push will pass PEP 8.

**Comment by EliAndrewC** [comment]
This is a deceptively difficult issue.  In general "show a more helpful message than a 500 error with a stacktrace" sounds easy, until you realize that we want to make a change to how the application behaves globally.

If this is still something an ambitious coding intern wants to tackle, then stepping back for a minute, I can definitely see why you chose to try putting this check in ``Tracking.track_pageview``, because that's a function which tends to get called whenever we pull up an Attendee record.  However, it's probably not the best place for this check, because there are places where it's NOT called where we still probably want to be doing this check.

Pretty much 100% of the time that we grab an ``Attendee`` object from the database by its id, we use the ``session.attendee(id)`` method.  This is a generated method whose code is at
https://github.com/magfest/ubersystem/blob/master/uber/models.py#L1818
and the specific calls in question are at
https://github.com/magfest/ubersystem/blob/master/uber/models.py#L1823
and
https://github.com/magfest/ubersystem/blob/master/uber/models.py#L1830

The behavior we want is to do something like

```python
try:
    self.query(model).filter_by(id=whatever).one()
except:
    self.rollback()
    raise HTTPRedirect('../wherever/custom_error_page?message={}', 'Maybe a message here with the model and id so the user knows what happened')
```

The ``rollback`` is important, because by default our codebase interprets ``HTTPRedirect`` as "everything went okay, so go ahead and do a commit before performing the redirect", which is probably NOT something we want in this case.

If we did this, then we'd be handling the "invalid id" case for ALL queries, and if we include both the ``model`` and the ``id`` in the error message, people will have useful information to pass along to the ``code@magfest.org`` mailing list.

**Comment by kitsuta** [comment]
It seems like it would be better to piggyback off our existing `_attendee_validity_check` function, which is here: https://github.com/magfest/ubersystem/blob/master/uber/models.py#L1910

The function currently only cares about whether a badge has an "Invalid" status, but there is no reason it can't also check if the badge _exists_ and simply return a different error page.

Eli may correct me on this though - it's possible that the 500 error happens before we get to that point.

**Comment by EliAndrewC** [comment]
Hmm, in theory you could put the try/except around the ``attendee = orig_getter(...)`` line.  Either way it's the same implementation, though if you put it there it would ONLY apply to ``Attendee`` model queries.  I'm not entirely sure whether that's a good or bad thing; I'm somewhat inclined to think we should just perform the check for all models.

**Comment by kitsuta** [comment]
@EliAndrewC Yeah, that's fair. I see no reason why we can't be nice to admins no matter what they're trying to access.

**Comment by kitsuta** [comment]
Hey, I'm checking all our PRs and I _just_ noticed that this was changed the day after I last commented. @timrxd in the future you definitely want to comment when you've made major changes for a PR - we don't get notified of commits.

I'll comment on the changes themselves in-line.

File: uber/models.py [hunk 3]
```diff
@@ -1805,12 +1805,18 @@ def track_pageview(cls, url, query):
             if "registration" in url:
                 with Session() as session:
                     attendee = session.query(Attendee).filter(Attendee.id == params['id']).first()
-                    Tracking.track(c.PAGE_VIEWED, attendee)
+                    if attendee is None:
```
Comment (kitsuta) [inline uber/models.py]: I believe this should be "if not attendee". That is more a style change than anything though.

File: uber/models.py [hunk 4]
```diff
@@ -1805,12 +1805,18 @@ def track_pageview(cls, url, query):
...
+                        return
+                    else:
+                        Tracking.track(c.PAGE_VIEWED, attendee)
             # Looking at a group's details
             elif "groups" in url:
                 with Session() as session:
                     group = session.query(Group).filter(Group.id == params['id']).first()
-                    Tracking.track(c.PAGE_VIEWED, group)
+                    if group is None:
```
Comment (kitsuta) [inline uber/models.py]: Same here as with my `if not attendee` comment above - `if not group` is more Pythonic, I believe.

File: uber/models.py [hunk 5]
```diff
@@ -1827,7 +1833,11 @@ def getter(self, params=None, *, bools=(), checkgroups=(), allowed=(), restricte
             if id == 'None':
                 inst = model()
             else:
-                inst = self.query(model).filter_by(id=id).one()
+                try:
+                    inst = self.query(model).filter_by(id=id).one()
+                except NoResultFound:
+                    self.rollback()
+                    raise HTTPRedirect('../preregistration/error_attendee_not_found?id={}', id)
```
Comment (kitsuta) [inline uber/models.py]: If I understand correctly, `getter()` is used to access ANY model - not just attendees. This should be a more generic page with a more generic message, if possible.

File: uber/site_sections/preregistration.py [hunk 6]
```diff
@@ -547,3 +547,6 @@ def shirt_reorder(self, session, message='', **params):
             'message': message,
             'attendee': attendee
         }
+
+    def error_attendee_not_found(self, id=''):
+        return "Sorry, but an attendee with badge ID %s doesn't exist." % id
```
Comment (kitsuta) [inline uber/site_sections/preregistration.py]: Uh... does this actually work? I would not expect this to work, or if it does, it wouldn't have the admin menu or anything like that. You should definitely use a template - you'd need to create one under templates/preregistration, and it needs the same name as this function (so right now it would be error_attendee_not_found.html, though I recommend change the name of the function and the template to something like model_not_found.html)

At any rate, you can improve this a bit by returning the name of the model (since, as I mentioned in the comment above, it is not just attendees anymore). Then in the template you can have a message like "Sorry, but the {{ model_name }} you tried to access does not exist." I do not know off the top of my head how to get the name of the model, but it _probably_ isn't too difficult. I hope.

**Comment by kitsuta** [comment]
Also I just noticed that all the files you edited changed from "100644" permissions to "100755" permissions. We... probably don't want this. @binary1230 may know more about how that works, though.

**Comment by binary1230** [comment]
we should fix the file permissions though it's likely that can only be done from a non-windows box since windows file permissions + git are broken.  good catch.

**Comment by timrxd** [comment]
Hey @kitsuta, yeah this was the first issue I worked on when I first started.  I wasn't very familiar with the codebase and this solution was definitely wrong.  I think we decided the rabbithole was too deep and complicated to go down at the time and I moved to a different issue.  This pull request should probably just be deleted.
